Psionic Eruption

TehFlaminTaco opened the pull request:
<!-- If you are new to the Delta-V repository, please read the [Contributing Guidelines](https://github.com/DeltaV-Station/Delta-v/blob/master/CONTRIBUTING.md) -->

## About the PR
<!-- What did you change? -->
Added **Psionic Eruption**, a power which detonates the user.

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
As of writing, all Psionic Powers are purely beneficial, and Psionic Manti rarely have any reason to 'cure' Psions. The added ability gives both the user and station crew a reason to be wearisome of psionics and treat them with the care and respect the lore commands of them.

## Technical details
<!-- Summary of code changes for easier review. -->
Adds a single new Psionic Power with 0.1 weight, **Psionic Eruption.**
When used, the Psion begins to shake violently, sparking, before exploding.
They can be stopped during this by standard stun or dispel behaviours (Such as pushing them or hitting them hard enough).
It typically takes 10 seconds for the Psion to detonate.

Possessing the ability causes periodical annoy messages for the Psion, prompting them to not forget about it. Additionally, a small amount of glimmer is generated each time. The frequency of these messages is determined by the Glimmer Tier.

The size of the explosion scales with Glimmer Tier, scaling from Media 3 at minimal, to Media 4 at dangerous.
Should the Glimmer Tier reach Critical (>900) however, the behaviour changes.
 * The detonate time increases to 30 seconds
 * The sparking whilst charging may shoot lightning
 * Explosion size as per Media 5

Glimmer reaching above 900 is rare enough that this should not be disruptive.

## Media

1.  <img width="660" height="369" alt="notready" src="https://github.com/user-attachments/assets/ba6d42b7-8ef6-48ce-95c9-03ef4da06a71" />
2.  <img width="99" height="89" alt="icon" src="https://github.com/user-attachments/assets/c4af803f-98cb-452e-9dba-d18fec058db2" />
3.  <img width="409" height="327" alt="smallboom" src="https://github.com/user-attachments/assets/7bcf96f2-9f67-4249-8471-1c7e39d5b16b" />
4.  <img width="495" height="359" alt="biggerboom" src="https://github.com/user-attachments/assets/6e056a5a-be4a-4795-8d82-a0b4eeb73385" />
5.  <img width="831" height="847" alt="image" src="https://github.com/user-attachments/assets/cdc1b24a-5b9a-4edd-8e65-0f2f2ba730ce" />





## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have tested all added content and changes.
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them. -->
N/A

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelogs must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl:
- add: Added a new power, Psionic Eruption

TehFlaminTaco committed:
Psionic Eruption

TehFlaminTaco committed:
Art Pass

YeenJunk commented:
This is beautiful

deltanedas reviewed (changes requested):
use file scoped namespaces
_DV not nyano folders
actually use _DV namespaces in classes

theres also no rule justification allowing a nonantag to nuke the station

Proxysseia commented:
direction review 24h

TehFlaminTaco committed:
Detangles spaghetti

Removed from Nyano namespaces, moved entirely into _DV namespaces and folders where possible.

TehFlaminTaco commented:
> use file scoped namespaces _DV not nyano folders actually use _DV namespaces in classes
> 
> theres also no rule justification allowing a nonantag to nuke the station

Corrected namespaces, I appreciate the style assistance.

I agree nuking as a nontag is generally unacceptable. The design intent was a gross-negligence disaster similar to a tesla going loose, in which the caster is a somewhat unintentional victim. I am happy to workshop this design more, or even just disable this extreme behaviour indefinitely.

AveriV2 commented:
DAGD with this finna go insane

Proxysseia reviewed (changes requested):
Decrease the size of the explosion at 900+ glimmer. Rest is good.

TehFlaminTaco committed:
Remove nuke behavior (Maybe next time 😢 )

Proxysseia commented:
Adding new psyonics is always cool, especially since psyonics are supposed to be kinda dangerous, having one that just blows you up is fun and funny. This opens up some unique RP with the noösphere and we are exited to see how it ends up being used in game.
Direction approved

TehFlaminTaco commented:
@deltanedas Would it be possible to re-review this? Let me know if anything else needs to be changed.

deltanedas commented on Content.Server/_DV/Abilities/Psionics/PsionicEruption/PsionicEruptionComponent.cs:
```diff
@@ -0,0 +1,33 @@
+    public TimeSpan NextAnnoy = TimeSpan.FromSeconds(5);
+
+    [DataField("warned")]
+    public bool Warned = false;
```
just use `[DataField]` and EntProtoId instead of string

do not store a InstantActionComponent no reason to do that

deltanedas commented on Content.Server/_DV/Abilities/Psionics/PsionicEruptionSystem.cs:
```diff
@@ -0,0 +1,240 @@
+        SubscribeLocalEvent<PsionicEruptionPowerComponent, PsionicEruptionDoAfterEvent>(OnDoAfter);
+    }
+
+    private void OnInit(EntityUid uid, PsionicEruptionPowerComponent component, ComponentInit args)
```
use Entity for all subscriptions this isnt nyano

deltanedas commented on Content.Server/_DV/Abilities/Psionics/PsionicEruptionSystem.cs:
```diff
@@ -0,0 +1,240 @@
+                    _jitteringSystem.DoJitter(uid, TimeSpan.FromSeconds(5), true, 10, 32);
+                    Spawn("EffectSparks", Transform(uid).Coordinates);
+                }, // it gibs, damage doesn't need to be high.
+                CancellationToken.None);
```
use TimeSpan for logic not Timer

deltanedas commented on Content.Server/_DV/Abilities/Psionics/PsionicEruptionSystem.cs:
```diff
@@ -0,0 +1,240 @@
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly EuiManager _euiManager = default!;
+    [Dependency] private readonly IPlayerManager _playerManager = default!;
+    [Dependency] private readonly MindSystem _mindSystem = default!;
```
sort and remove System and Manager from field names

deltanedas commented on Content.Server/_DV/Abilities/Psionics/PsionicEruptionSystem.cs:
```diff
@@ -0,0 +1,240 @@
+        var pos = _transformSystem.GetMapCoordinates(uid);
+        _body.GibBody(uid, acidify: true, body, launchGibs: true);
+        int boom = 4;
+        switch (_glimmer.GetGlimmerTier(_glimmer.Glimmer))
```
use switch expression

deltanedas commented on Resources/Prototypes/Nyanotrasen/psionicPowers.yml:
```diff
@@ -10,3 +10,4 @@
     NoosphericZapPower: 0.3
 #    PsionicInvisibilityPower: 0.15
     MindSwapPower: 0.15
+    PsionicEruptionPower: 0.1 # DeltaV
```
dont need comment

deltanedas commented on Resources/Textures/_DV/Interface/Actions/actions_psionics.rsi/meta.json:
```diff
@@ -42,6 +42,9 @@
             "name": "telegnosis"
+        },
+        {
+            "name": "eruption"
```
update copyright

deltanedas reviewed (changes requested):
change the eui to use xaml too

TehFlaminTaco committed:
Various maintainer requested patches

TehFlaminTaco committed:
Merge branch 'master' into psionic-eruption

Signed-off-by: William Lemon <William.Lemon2@gmail.com>

TehFlaminTaco commented:
@deltanedas Cleaned up following requested changes. Only issue I had was changing from `Eui`, upon review other systems seem to still use EUI to open the window on the client. (Using coscult as a reference). If there is a better way to do that, please let me know.

deltanedas commented:
euis can still use xaml see the eorg popup

TehFlaminTaco commented:
I now use EUI to show the XAML, I just wasn’t sure if there was a preferred
alternative method

On Wed, 30 Jul 2025 at 20:59, deltanedas ***@***.***> wrote:

> *deltanedas* left a comment (DeltaV-Station/Delta-v#4078)
> <https://github.com/DeltaV-Station/Delta-v/pull/4078#issuecomment-3135805627>
>
> euis can still use xaml see the eorg popup
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DeltaV-Station/Delta-v/pull/4078#issuecomment-3135805627>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACAVREHCSNY4W75QFL43PZL3LCQQPAVCNFSM6AAAAACCHX2E4CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMZVHAYDKNRSG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

deltanedas commented on Content.Client/_DV/CosmicCult/UI/CosmicDeconvertedMenu.xaml.cs:
```diff
@@ -1,3 +1,4 @@
+
```
untouch

deltanedas commented on Content.Server/_DV/Abilities/Psionics/PsionicEruption/PsionicEruptionSystem.cs:
```diff
@@ -0,0 +1,226 @@
+                        _lightning.ShootRandomLightnings(uid, 5f, _random.Next(1, 3), "Lightning", 0, false);
+                    }
+                    _jittering.DoJitter(uid, TimeSpan.FromSeconds(5), true, 10, 32);
+                    Spawn("EffectSparks", Transform(uid).Coordinates);
```
store this as a static readonly EntProtoId on the system

TehFlaminTaco committed:
More cleanup

TehFlaminTaco committed:
Merge branch 'psionic-eruption' of https://github.com/TehFlaminTaco/Delta-v into psionic-eruption

TehFlaminTaco commented:
@deltanedas All done. Hopefully shouldn't keep you busy for much longer.

deltanedas commented:
theres still like 6 more EffectSparks

TehFlaminTaco committed:
I'm blind

TehFlaminTaco commented:
@deltanedas should now be good to go. Feel free to smite me if I missed one
